src/buffer.rs

```diff
@@ -1,4 +1,19 @@
-//! Contains the generic `ImageBuffer` struct.
+//! Defines standard containers for image data.
+//!
+//! The best container depends on the use:
+//! * The generic [`ImageBuffer`] struct, a simplified pixel container with a standard matrix
+//!   layout of uniform channels. The strength is convenient access for modifications with a direct
+//!   mapping and access to the underlying data.
+//! * The [`DynamicImage`] enumeration builds on `ImageBuffer` to provide a standard selection of
+//!   basic channel representations, and conventionally in sRGB color space. This is usually enough
+//!   to represent any source data without loosing much precision. This makes is suitable for
+//!   basic, but generic, image operations.
+//! * The [`Canvas`] struct as general texel container of general layouts as a buffer. The strength
+//!   is the generality (and extensibility) that allows its use for interchange of image data and
+//!   reduced processing cost if the convenience above is not required.
```

Comment:
Is this somewhat clear motivation?

Reply:
Sorry for the delay responding. That explains what sorts of images the different types can store, but doesn't really think through what the user can _do_ with them. My rough thinking:

* `ImageBuffer` provides both a collection of image processing methods (resize, crop, etc.) as well as ergonomic pixel getters and setters to implement your own image processing. The serious drawback is that the color type is determined at compile time. One consequence is that saving works fine but loading files as ImageBuffers isn't directly supported.
* `DynamicImage` supports even more built in image processing as well as loading/saving a whole range of file formats. Conversion between color types is supported. The enum is non-exhaustive but users can implement their own algorithms by matching on the current variants with special case logic for each one.
* `Canvas` holds and converts between a broader range of color types. No other image processing algorithms are provided and users who want to write their own must... rely on raw access to aligned byte slices? Cast them to FlatSamples first? That's perhaps a bit unfair of a characterization since the type could still be useful as a common "interchange" representation for images data in Rust. However, I suspect it would have a better chance of catching on if users didn't have to immediately convert the Canvas into something else to operate on it.

Reply:
> No other image processing algorithms are provided and users who want to write their own must... rely on raw access to aligned byte slices?

Precisely. The best reason to use the buffer is uploading it to a GPU texture as directly as possible; for which only the layout and raw bytes are necessary and sufficient, and all further CPU processing is less precise and more costly than necessary. (And I stopped expanding the interface, before anyone could get the idea of making this PR contingent on a corresponding `wgpu` GPU framework).

The lack of direct operations is by design. Precisely because we may not have interpreted the correct color information (color space, or ICC, or otherwise) yet, just the need to store some numeric texel matrix anyways. There's just nothing sensible to do with the matrix entries, at that point.  It should get support for Yuv layouts and CMYK rather soon to fix the tragedy of wrongly applying incorrect ICC tables while loading jpeg, WebP, png and tiff, as we are doing currently... But that's for another PR in my opinion.